src/Core/Content/Product/SalesChannel/Detail/ProductConfiguratorLoader.php

```diff
@@ -209,7 +209,7 @@ private function buildCurrentOptions(SalesChannelProductEntity $product, Propert
         $keyMap = $groups->getOptionIdMap();
 
         $current = [];
-        foreach ($product->getOptionIds() as $optionId) {
+        foreach ($product->getOptionIds() ?? [] as $optionId) {
```

Comment:
why should this happen?
This function is only called with a variant product. How can it happen that the variant has no option ids?
That would be a broken variant

Reply:
Is there some kind of validation I can extend to prevent a variant from being activated if there are no options selected? That would help my case more than making it broken in a nicer way :)

Reply:
if you generate variant over the admin, it is not possible that the variant is generated without options. 

If you use the api, yeah that is possible, but we have no validation for this case. This is simply data that must be present.

But letting the configurator loaded without the current combination doesn't make the page very usable, because namely the current selection is not set in the selects.

This in turn would only lead to multiple of other consequential errors, which we would then have to handle separately.

And validating this data every time a variant is written via the API would be at the expense of performance. This would also "punish" the write requests that send the data correctly.

Reply:
So I would recommend throwing an exception here and then displaying an appropriate error page. The corresponding product number can be written in a log.

Reply:
I disagree for multiple reasons:
- If data is invalid we shouldn't accept it
- checking wether an array is present and not empty if `parent_id != null` isn't a bad performance impact (at least not in my world 😅)
- Showing an error to a customer (we never hear of) is horrible compared to an error one of our developers see
- and a personal one: currently another department is writing broken code I need to handle, having a validation makes sure that we fix the source of the error and not trying to hack it (as and @johmanx10 did - with due respect :D)

if someone tells me how/where to implement a validation I'm happy to do it!
I can't find any validation here - most likely I'm not deep enough yet.
<img width="1010" alt="image" src="https://user-images.githubusercontent.com/379680/121768027-331e8500-cb5c-11eb-8493-84c37e772bad.png">

Reply:
Good point made by @t2oh4e https://twitter.com/_t2oh4e_/status/1403611181588389891 ... we don't have the variant data all the time with us. What we can do at least is check on creation. I think that would already help. But then we have inconsistent behaviour. I'm not sure about this - but having broken variants drives me crazy 😓

Reply:
> checking wether an array is present and not empty if parent_id != null isn't a bad performance impact (at least not in my world )

If it was done with a simple parent_id !== null, it would not be a performance impact, not even in my world don't worry.
But if you want to do it right, you have to check:
- that a parent product should have no options
- if an option is deleted, the variant is not without options
- if an option is deleted also the configurator is updated

The only problem is that it would be valid to do this in several write-requests. Therefore the question - When should one validate something like that?

> Showing an error to a customer (we never hear of) is horrible compared to an error one of our developers see

Of course, the error "option ids is not filled, invalid variant" should not be displayed, but a customer-friendly 404 page.

**I would now suggest the following solution:**

1. `ProductConfiguratorLoader.php` throw an exception if option Ids is not set.

2. set `options` association in `ProductDefinition` to required, but the whole thing must be behind a feature flag because it is a breaking change. You will also have many failling tests.